pkg/reconciler/taskrun/validate_resources.go

```diff
@@ -122,6 +122,59 @@ func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params
 		return fmt.Errorf("param types don't match the user-specified type: %s", wrongTypeParamNames)
 	}
 
+	return ValidateObjectParameters(paramSpecs, params)
+}
+
+// ValidateObjectParameters checks if all required keys of object type params are provided in taskrun params.
+// To use this function externally, make sure `paramSpecs` and `params` have
+// been checked against general missing/extra params and param type alignment,
+// which been done in `validateParams` in this module.
+func ValidateObjectParameters(paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) error {
```

Comment:
Same logic applies here what if somewhere else wants to validate against one single param instead of the whole param list? I'm not sure if your module is providing that flexibility. You should put the loops in `validateParams`, and have:
```
ValidateObjectParameters(ps *v1beta1.ParamSpec, p *v1beta1.Param) {

}
...
func validateParams(ctx context.Context, paramSpecs []v1beta1.ParamSpec, params []v1beta1.Param) error {
... 
missedkeys := []
for _, spec := range paramSpecs {
		if spec.Type == v1beta1.ParamTypeObject {
			m = ValidateObjectParameters()
                         if len(m) != 0 {
                           missedkeys.append(m)
                         }
		}
	}
}
// error for all params together

Reply:
Ok, I see the point you're trying to make.

But wouldn't that be too specific? I don't see any scenario in the current code base which is to validate against one single param. I might miss that if there is one, or if you have some scenarios in mind which might need to validate against a single case in future, can you please explain a bit more?

More importantly, to support what you just mentioned - validate against a single item, do you think the existing functions like [validateParams](https://github.com/tektoncd/pipeline/blob/30bed9f1c4fb834baf1cf12fd531b472821a3ce1/pkg/reconciler/taskrun/validate_resources.go#L72-L126), [validateResources](https://github.com/tektoncd/pipeline/blob/30bed9f1c4fb834baf1cf12fd531b472821a3ce1/pkg/reconciler/taskrun/validate_resources.go#L33-L70) should all be changed as well ?

Reply:
Just remember that @wlynch also [commented](https://github.com/tektoncd/pipeline/pull/4786#discussion_r863916027) before that we want to extract all object-specific validation out to not interleave and avoid context switching for readers. I'll let @wlynch to chime in to get more inputs on this :)

Reply:
> 

One use case is that in Chains, we probably wouldn't match each params or objects in groups. They could be matched against each other. When you have module with single item, it's super easy to extend another module to validate against all the params together, and that way it's super easy to wrap those loops within another layer of module. I don't see the problem there

Reply:
> But wouldn't that be too specific? 

I'm not sure about this is being too specific, as this is just providing a flexible module that can be easily extended when we want another level of aggregation.

Reply:
@dibyom @vdemeester @pritidesai Hi team we probably need more thoughts about making these as sharable modules.

Reply:
update: just update the code to align with the refactor for `validateParams` function that @jerop 's made in #4814. 

Can you please have a look at the [objectParam-specific functions](https://github.com/tektoncd/pipeline/pull/4786/files#diff-f6ff4d26a45baec7685e7ad851e9d7fa5901e719b4635d2c7de0298d08b8a71dR150-R198) which is just added and is also part of `validateParams`? @jerop Let me know what you think. Thanks!

Reply:
My 2c - I'm a little skeptical if we would actually need to validate params again in chains? If we can validate that a message came from the controller via spire, would there be a need to re-validate param fields? lmk off-thread if you have something else in mind - would be happy to chat more about this. 👀 

Out of scope of this PR, but in general I think we may be exporting too much internal validation out through the package? e.g. idk if anyone besides the controller would call something like [`ValidateParameterTypes`](https://pkg.go.dev/github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1#ValidateParameterTypes) - I'd expect most users that care about validation to just call through the Validatable interfaces on the CRD types.

For this PR - let's try and unexport as much as possible. This gives us flexibility to make changes with how validation is structured without needing to worry about breaking downstream clients (i.e. we can rewrite this later if we want).

Reply:
> For this PR - let's try and unexport as much as possible. This gives us flexibility to make changes with how validation is structured without needing to worry about breaking downstream clients (i.e. we can rewrite this later if we want).

Done. Just unexported them.

Agree. Let's leave the function signature as is to align with [other param-related validation functions](https://github.com/tektoncd/pipeline/blob/e586418e798ae9fcc5193d1ed09b810d912bca1b/pkg/reconciler/taskrun/validate_resources.go#L107-L145). We can rewrite it later if needed.

Reply:
> would there be a need to re-validate param fields? lmk off-thread if you have something else in mind - would be happy to chat more about this. 👀
> 
> Out of scope of this PR, but in general I think we may be exporting too much internal validation out through the package? e.g. idk if anyone besides the controller would call somethin

Can you elaborate a bit more about validating via spire? I think in Tekton Chains, when we have a defined schema, we still need to validate the objects against the defined schema.

I think it makes sense to export the function here, as at least in results there is a clear need to for such sharable module. I do agree that we can rewrite this piece of code later, but when I see a clear need of a piece of code, I think it worths thinking and discussing  about it now rather than later.